### PR TITLE
Persistent node positions and first steps of creating SDFG elements graphically

### PR DESCRIPTION
**Pull request by tobiasae**
Contains changes for persistent node positions, improved search GUI and preliminary work for adding SDFG elements graphically

### CODE REVIEW

File: sdfg_utils.js [hunk 1]
```diff
@@ -181,19 +323,18 @@ function parse_sdfg(sdfg_json) {
     return JSON.parse(sdfg_json, reviver);
 }
 
-function isDict(v) {
-    return typeof v === 'object' && v !== null && !(v instanceof Array) && !(v instanceof Date);
-}
+function replacer(name, val) {
```
Comment (phschaad) [inline sdfg_utils.js:326]: What was the rationale behind changing the replacer? Did it fail on circular object structures?

File: renderer.js [hunk 2]
```diff
@@ -1249,8 +1340,15 @@ class SDFGRenderer {
         // Toolbar-related fields
         this.menu = null;
         this.toolbar = null;
+        this.panmode_btn = null;
         this.movemode_btn = null;
         this.selectmode_btn = null;
+        this.addmode_btns = null;
+        this.add_type = null;
+        this.addmode_divs = null;
+        this.mode_selected_bg_color = "#CCCCCC";
```
Comment (phschaad) [inline renderer.js:1349]: Nitpick: The selection color is a lot better the way you do it in context of the VSCode extension - the bright gray is a bit hard to read for standalone SDFV at least, imo :-)

File: renderer.js [hunk 3]
```diff
@@ -1486,38 +1617,88 @@ class SDFGRenderer {
...
+            for (let add_btn of this.addmode_btns) {
+                add_btn.onclick = () => {
+                    this.mouse_mode = 'add';
+                    this.add_type = add_btn.getAttribute('type');
+                    this.update_toggle_buttons();
+                };
+            }
+            this.mode_selected_bg_color = '#22A4FE';
+        } else {
+            // Mode buttons are empty in plain webview
```
Comment (phschaad) [inline renderer.js:1635]:
```suggestion
            // Mode buttons are empty in standalone SDFV
```

File: renderer.js [hunk 4]
```diff
@@ -2479,7 +2934,25 @@ class SDFGRenderer {
                     dirty = true;
                     element_focus_changed = true;
                 }
+                if (this.mouse_mode === 'move') {
+                    // Send new SDFG to vscode extension
```
Comment (phschaad) [inline renderer.js:2938]:
```suggestion
```

Comment (phschaad) [reply renderer.js:2938]: Redundant

File: renderer.js [hunk 5]
```diff
@@ -1695,6 +2059,10 @@ class SDFGRenderer {
             if ('is_collapsed' in obj.attributes && !obj.type.endsWith('Exit'))
                 obj.attributes.is_collapsed = true;
         });
+
+        // Send updated SDFG to vscode extension
+        this.send_new_sdfg_to_vscode();
```
Comment (phschaad) [inline renderer.js:2064]: Why do we do this when zooming to a view - or rather: when can the SDFG change based solely on a `zoom_to_view` action?

File: renderer.js [hunk 6]
```diff
@@ -1601,6 +1782,75 @@ class SDFGRenderer {
         if (user_transform === null)
             this.zoom_to_view();
 
+        let svgs = {}
```
Comment (phschaad) [inline renderer.js:1785]:
```suggestion
        let svgs = {};
```

**Review by phschaad — CHANGES_REQUESTED** [review]
Very nice work! The new interaction with the mode buttons - particularly using the keyboard modifiers - seems intuitive and more user friendly.
Only a few minor comments/nitpicks and questions.

**Comment by phschaad** [comment]
I went ahead and addressed the comments I had left and moved this to a new PR (#31). You of course remain the author of all your commits :-)
I'm closing this PR as a result.

File: sdfg_utils.js (hunk 1 above)
Comment (tobiasae) [reply sdfg_utils.js:326]: Yes, the main reason was to prevent adding Edge objects to break the circular structure.

**Comment by tobiasae** [comment]
Great! Thanks a lot :)

**Pull request closed without merging**